File: lib/services/external_launcher.dart
import 'dart:io';
import 'package:url_launcher/url_launcher.dart' as launcher;
import '../utils/constants.dart';

class ExternalLauncher {
  static Future<void> openOutlook({String? email, String? emailId}) async {
    // Prefer opening a specific mail by searching the id if provided
    if (emailId != null && emailId.isNotEmpty) {
      final uri = Uri.parse('outlook:search?text=$emailId');
      if (await launcher.canLaunchUrl(uri)) {
        await launcher.launchUrl(uri);
        return;
      }
    }

    // Try mailto scheme (default client is usually Outlook on Windows)
    if (email != null && email.isNotEmpty) {
      final uri = Uri(scheme: 'mailto', path: email);
      if (await launcher.canLaunchUrl(uri)) {
        await launcher.launchUrl(uri);
        return;
      }
    } else {
      final uri = Uri(scheme: 'mailto');
      if (await launcher.canLaunchUrl(uri)) {
        await launcher.launchUrl(uri);
        return;
      }
    }

    // Windows fallback: configured Outlook path or outlook: protocol
    if (Platform.isWindows) {
      try {
        final path = AppConstants.outlookPathWindows;
        if (path != null && path.isNotEmpty) {
          await Process.start(path, [], runInShell: true);
          return;
        }
        final uri = Uri.parse('outlook:');
        if (await launcher.canLaunchUrl(uri)) {
          await launcher.launchUrl(uri);
          return;
        }
      } catch (_) {}
    }

    // macOS fallback
    if (Platform.isMacOS) {
      try {
        await Process.run('open', ['-a', 'Microsoft Outlook']);
      } catch (_) {}
    }
  }

  static Future<void> openDingTalk() async {
    // Prefer scheme
    try {
      final uri = Uri.parse('dingtalk://dingtalkclient/action/open');
      if (await launcher.canLaunchUrl(uri)) {
        await launcher.launchUrl(uri);
        return;
      }
    } catch (_) {}

    if (Platform.isWindows) {
      try {
        final path = AppConstants.dingTalkPathWindows;
        if (path != null && path.isNotEmpty) {
          await Process.start(path, [], runInShell: true);
          return;
        }
        await Process.run('cmd', ['/c', 'start', 'DingTalk'], runInShell: true);
        return;
      } catch (_) {}
    }

    if (Platform.isMacOS) {
      try {
        await Process.run('open', ['-a', 'DingTalk']);
        return;
      } catch (_) {}
    }
  }
}

File: src/ChatDesktop.Infrastructure/External/ExternalLauncher.cs
using System.Diagnostics;
using ChatDesktop.Infrastructure.Config;

namespace ChatDesktop.Infrastructure.External;

/// <summary>
/// 外部应用启动
/// </summary>
public static class ExternalLauncher
{
    public static void OpenOutlook(string? email, string? emailId)
    {
        if (!string.IsNullOrWhiteSpace(emailId))
        {
            TryOpenUri($"outlook:search?text={emailId}");
        }

        if (!string.IsNullOrWhiteSpace(email))
        {
            TryOpenUri($"mailto:{email}");
        }
        else
        {
            TryOpenUri("mailto:");
        }

        if (OperatingSystem.IsWindows())
        {
            var path = EnvConfig.OutlookPathWindows;
            if (!string.IsNullOrWhiteSpace(path))
            {
                TryStartProcess(path);
                return;
            }

            TryOpenUri("outlook:");
        }
        else if (OperatingSystem.IsMacOS())
        {
            TryStartProcess("open", "-a", "Microsoft Outlook");
        }
    }

    public static void OpenDingTalk()
    {
        TryOpenUri("dingtalk://dingtalkclient/action/open");

        if (OperatingSystem.IsWindows())
        {
            var path = EnvConfig.DingTalkPathWindows;
            if (!string.IsNullOrWhiteSpace(path))
            {
                TryStartProcess(path);
                return;
            }

            TryStartProcess("cmd", "/c", "start", "DingTalk");
        }
        else if (OperatingSystem.IsMacOS())
        {
            TryStartProcess("open", "-a", "DingTalk");
        }
    }

    private static void TryOpenUri(string uri)
    {
        try
        {
            var psi = new ProcessStartInfo
            {
                FileName = uri,
                UseShellExecute = true,
            };
            Process.Start(psi);
        }
        catch
        {
        }
    }

    private static void TryStartProcess(params string[] args)
    {
        try
        {
            if (args.Length == 0)
            {
                return;
            }

            var psi = new ProcessStartInfo
            {
                FileName = args[0],
                UseShellExecute = true,
            };

            if (args.Length > 1)
            {
                psi.ArgumentList.AddRange(args.Skip(1));
            }

            Process.Start(psi);
        }
        catch
        {
        }
    }
}
